### ISSUE TITLE
[Feature] Add support for tuple destructuring 

### ISSUE THREAD

**Issue by CC007** [manifold-systems/manifold#485]
**Is your feature request related to a problem? Please describe.**
Consider the following method:
```java
auto returnTuple() {
    return "foo", 42, 13.37, (LocalDateTime.now());
}
```
When you want to use specific components of this tuple, it becomes quite verbose:

```java
var tupleItems = returnTuple();

// You'll have to use the item1, item2, etc. to access the items
String s = tupleItems.item1;
int i = tupleItems.item2;
double d = tupleItems.item3;
LocalDateTime ldt = tupleItems.item4;
System.out.println("s = $s, i = $i, d = $d, ldt = $ldt");

// or directly, which makes the line a bit longer
System.out.println("s = ${tupleItems.item1}, i = ${tupleItems.item2}, d = ${tupleItems.item3}, ldt = ${tupleItems.item4}");
```

**Describe the solution you'd like**
It would be nice if manifold supported tuple destructuring, like so: 
```java
var (s, i, d, ldt) = returnTuple();
System.out.println("s = $s, i = $i, d = $d, ldt = $ldt");
```
This would also be a nice feature for interacting with records, when you want to more formally define and document the data structure. Consider this:
```java
/**
 * This describes a person
 * 
 * @param name name of person
 * @param age age of person
 */
record Person(
    String name, 
    int age
) {}

...

String doSomethingWithPerson(PersonSupplier personSupplier) {
    (name, age) = personSupplier.getPerson();
    return "$name is $age years old";
}
```

**Describe alternatives you've considered**
The syntax could also be like this for when you don't want to use type inference:
```java
(String s, int i, double d, LocalDateTime ldt) = returnTuple();
```

**Additional context**
See documentation for tuple deconstructing/destructuring  in other languages:
- C#: https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/functional/deconstruct
- Kotlin: https://kotlinlang.org/docs/destructuring-declarations.html

*PS: why are the extra brackets needed around `LocalDateTime.now()`? It doesn't compile without them.*

**Comment by rsmckinney** [comment]
Hi @CC007. All great features requests.

Regarding, the LocalDateTime.now() question, I'm not seeing that.
```java
  auto returnTuple() {
    return "foo", 42, 13.37, LocalDateTime.now();
  }
```
Works fine for me, both in IJ editor & compilation.

**Comment by CC007** [comment]
I'm getting this:
![afbeelding](https://github.com/manifold-systems/manifold/assets/5381337/1f0a5c12-9bcc-4e62-87d0-b0273237236f)

So without brackets compiles alright, but the item isn't usable.

It got compiled to this instead:
![afbeelding](https://github.com/manifold-systems/manifold/assets/5381337/1326686b-4179-49d9-8482-f41b4cd07a78)

**Comment by rsmckinney** [comment]
Right. That is intentional. Tuples infer labels from the expressions if possible. In this case the method name.

**Comment by CC007** [comment]
It is a bit unintuitive though that item4 couldn't be used as well as an alias.

**Comment by rsmckinney** [comment]
> It is a bit unintuitive though

Yep. But I think label inference tends to be more useful than not because:

1. Code completion.<br>I would think the vast majority of manifold users are developing with IJ / Android where the tuple items are at your fingertips.

2. Explicit labels.<br>Unless the labels are apparent from the expressions (refs to fields, locals, properties, etc.) direct labels are best.

3. Low arity.<br>Most tuples consist of two items. If you’re using tuples, you probably like labeling your two items explicitly, otherwise you’d probably prefer a ‘Pair’ class or the like. Just my take on it.

**Comment by CC007** [comment]
Those are indeed all good reasons why label inference is intuitive to have. What I found unintuitive though was not that there is inference, but the lack of the item4. I'm a proponent of having both.

**Comment by rsmckinney** [comment]
Sure, that makes sense. Always have the `itemN` syntax regardless of the labels. I like it.

Would be interesting to make array operator work where a compile-time constant index expression would determine the type:

```java
String item1 = tuple[0];
int item2 = tuple[1];
```

lol

**Comment by CC007** [comment]
Agreed. Actually wanted to add that as a feature request too, but considering the type differences, I could see that being an issue.

**Comment by CC007** [comment]
Any progress on implementing this feature?

**Comment by rsmckinney** [comment]
The plan for supporting this feature was to add itemN properties (via manifold-props) for all the tuple components and also add properties for inferred labels. However, it was decided that bringing in the manifold-props dependency to manifold-tuple was not worth the extra load.

What complicates this feature (and why it sits here for years) is that, for consistency, it must be accessed with field syntax. But aside from properties, there is no other way to achieve that with manifold. I am considering making this a one-off field resolution condition where a tuple field reference to itemN, when not a field, is treated as an alias to the Nth field.
